templates/components/engines/typescript/agent/tools/interpreter.ts

```diff
@@ -0,0 +1,171 @@
+import { CodeInterpreter, Logs, Result } from "@e2b/code-interpreter";
+import type { JSONSchemaType } from "ajv";
+import fs from "fs";
+import { BaseTool, ToolMetadata } from "llamaindex";
+import crypto from "node:crypto";
+import path from "node:path";
+
+export type InterpreterParameter = {
+  code: string;
+};
+
+export type InterpreterToolParams = {
+  metadata?: ToolMetadata<JSONSchemaType<InterpreterParameter>>;
+  apiKey?: string;
+};
+
+export type InterpreterToolOuput = {
+  isError: boolean;
+  logs: Logs;
+  extraResult: InterpreterExtraResult[];
+};
+
+type InterpreterExtraType =
+  | "html"
+  | "markdown"
+  | "svg"
+  | "png"
+  | "jpeg"
+  | "pdf"
+  | "latex"
+  | "json"
+  | "javascript";
+
+export type InterpreterExtraResult = {
+  type: InterpreterExtraType;
+  // url: string;
+  filename: string;
+};
+
+const DEFAULT_META_DATA: ToolMetadata<JSONSchemaType<InterpreterParameter>> = {
+  name: "interpreter",
+  description: `
+  - You are a Python interpreter.
+  - You are given tasks to complete and you run python code to solve them.
+  - The python code runs in a Jupyter notebook. Every time you call \`interpreter\` tool, the python code is executed in a separate cell. It's okay to make multiple calls to \`interpreter\`.
+  - Display visualizations using matplotlib or any other visualization library directly in the notebook. Shouldn't save the visualizations to a file, just return the base64 encoded data.
+  - You can install any pip package (if it exists) if you need to but the usual packages for data analysis are already preinstalled.
+  - You can run any python code you want in a secure environment.
+  `, // TODO: Add more guide to help AI use data to generate code (eg. wiki tool, google sheet tool data)
+  parameters: {
+    type: "object",
+    properties: {
+      code: {
+        type: "string",
+        description: "The python code to execute in a single cell.",
+      },
+    },
+    required: ["code"],
+  },
+};
+
+export class InterpreterTool implements BaseTool<InterpreterParameter> {
+  private readonly outputDir = "data";
+  private readonly dataAPI = "/api/data/";
+  private apiKey?: string;
+  metadata: ToolMetadata<JSONSchemaType<InterpreterParameter>>;
+  codeInterpreter?: CodeInterpreter;
+
+  constructor(params?: InterpreterToolParams) {
```

Comment:
if an agent is using this tool, how often is the tool initiated? is it the lifetime of the task or of the agent?

Reply:
i think the problem is that the agent class is stateful. what happens if we reuse the agent instance between requests?
ideally, each user probably should have its own agent? let's deal with this issue in another PR and just test if it works with a codeinterpreter as is